Sources/ComposableArchitecture/Reducer.swift

```diff
@@ -189,6 +189,61 @@ public struct Reducer<State, Action, Environment> {
       return self.reducer(&state!, action, environment)
     }
   }
+    
+    /// Transforms a reducer that works on non-optional state into one that works on optional state by
+    /// only running the non-optional reducer when state is non-nil.
+    ///
+    /// Often used in tandem with `pullback` to transform a reducer on a non-optional local domain
+    /// into a reducer on a global domain that contains an optional local domain:
+    ///
+    ///     // Global domain that holds an optional local domain:
+    ///     struct AppState { var modal: ModalState? }
+    ///     struct AppAction { case modal(ModalAction) }
+    ///     struct AppEnvironment { var mainQueue: AnySchedulerOf<DispatchQueue> }
+    ///
+    ///     // A reducer that works on the non-optional local domain:
+    ///     let modalReducer = Reducer<ModalState, ModalAction, ModalEnvironment { ... }
+    ///
+    ///     // Pullback the local modal reducer so that it works on all of the app domain:
+    ///     let appReducer: Reducer<AppState, AppAction, AppEnvironment> =
+    ///       modalReducer.optional().pullback(
+    ///         state: \.modal,
+    ///         action: /AppAction.modal,
+    ///         environment: { ModalEnvironment(mainQueue: $0.mainQueue) }
+    ///       )
+    ///
+    /// - See also: `IfLetStore`, a SwiftUI helper for transforming a store on optional state into a
+    ///   store on non-optional state.
+    /// - See also: `Store.ifLet`, a UIKit helper for doing imperative work with a store on optional
+    ///   state.
+    public func optional(_ file: StaticString = #file, _ line: UInt = #line) -> Reducer<State?, Action, Environment> {
+      .init { state, action, environment in
+        guard state != nil else {
+          assertionFailure(
+            """
+            "\(debugCaseOutput(action))" was received by an optional reducer at \(file):\(line) \
```

Comment:
Assuming `file` and `line` make it into the console in a reasonable manner wanna change this messaging to be the same as the existing one?

Reply:
Yes that's what I was changing now but then I realised I don't see any difference passing file/line to the assert 🤔 I was expecting the error line to stop at the point of origin like what XCTest assertions do. 
![Screenshot 2020-07-08 at 19 30 35](https://user-images.githubusercontent.com/750912/86951255-8e2c4880-c151-11ea-8d0f-fc99b71a3144.png)

Reply:
Isn't it showing up at the very end here?

<img width="764" alt="image" src="https://user-images.githubusercontent.com/658/86956525-34af2400-c127-11ea-88f2-e28acdf85c62.png">

Reply:
And yeah it'd be nice for Xcode to trap on those lines instead, too :( Maybe feedback should be filed.

Reply:
You're right! I thought that was part of the message somehow. All good then 👍